### PR TITLE
Add stubbs laser; Update mock server; update documentation

### PR DESCRIPTION
**Pull request by couger01**

### CODE REVIEW

File: results.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+============================= test session starts ==============================
```
Comment (wvreeven) [inline results.txt:1]: Please remove this file.

File: python/lsst/ts/tunablelaser/mock_server.py [hunk 2]
```diff
@@ -29,10 +35,37 @@
...
+    Attributes
+    ----------
+    device : `MockNT252`
+        The mock NT252 device.
+    """
+
+    def __init__(self) -> None:
+        self.device = MockNT252()
+        super().__init__(
+            port=9001,
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/mock_server.py:53]: Please do not hardcode the values for port, host, encoding and terminator but take them from configuration.

File: python/lsst/ts/tunablelaser/mock_server.py [hunk 3]
```diff
@@ -41,7 +74,7 @@ class MockServer(tcpip.OneClientServer):
         The port that the server will start on.
     """
 
-    def __init__(self, port=0) -> None:
+    def __init__(self, port=9000) -> None:
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/mock_server.py:77]: Please do not hardcode the values for port, host, encoding and terminator but take them from configuration.

File: python/lsst/ts/tunablelaser/mock_server.py [hunk 4]
```diff
@@ -117,6 +131,152 @@ def __repr__(self):
...
+        The output energy level.
+    display_current : `int`
+        The display current.
+    burst_length : `int`
+        The burst length.
+    """
+
+    def __init__(self) -> None:
+        self.log = logging.getLogger(__name__)
+        self.wavelength = 650
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/mock_server.py:159]: Please do not hardcode the values for wavelength, temperature and display_current but use random generators with ranges fit for those values. See the ts_ess_common package for an example.

File: tests/test_ascii.py [hunk 5]
```diff
@@ -61,53 +63,55 @@ def test_create_set_message(self):
...
             self.bad_settable_ascii_register = AsciiRegister(
-                commander=None,
+                component=None,
                 module_name="Bad",
                 module_id=0,
                 register_name="Bad",
                 read_only=False,
             )
 
+    @pytest.mark.skip("Not working.")
```
Comment (wvreeven) [inline tests/test_ascii.py:73]: Please create a ticket to make the tests that are marked as `skip` work and add TODOs in this class.

File: .gitignore [hunk 6]
```diff
@@ -128,3 +128,4 @@ pytest_session.txt
 .mypy.ini
 .clang-format
 .ruff.toml
+towncrier.toml
```
Comment (wvreeven) [inline .gitignore:131]: You committed `towncrier.toml` so please remove this line.

File: python/lsst/ts/tunablelaser/ascii.py [hunk 7]
```diff
@@ -30,133 +30,10 @@
...
-        decoded_message : `str`
-            The decoded and boilerplate stripped string.
-        """
-        decoded_message = reply.decode("ascii").rstrip("nmC\r\n\x03")
-        if decoded_message.startswith("'''"):
-            raise Exception(decoded_message)
-        return decoded_message
-
 
 class AsciiRegister:
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/ascii.py:38]: Please rename the file to have a more descriptive name than `ascii.py`, for instance `ascii_register.py`.

File: python/lsst/ts/tunablelaser/ascii.py [hunk 8]
```diff
@@ -167,7 +44,7 @@ class AsciiRegister:
 
     Parameters
     ----------
-    commander : `TCPIPClient`
+    component : `Laser`
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/ascii.py:47]: Please update the description of this parameter to reflect what it actually is. It is not a TCP/IP stream.

File: python/lsst/ts/tunablelaser/hardware.py [hunk 9]
```diff
@@ -54,8 +55,8 @@ class CPU8000:
 
     Parameters
     ----------
-    commander : `TCPIPClient`
-        A reference to the tcp/ip client.
+    component : `Laser`
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/hardware.py:58]: This comment doesn't apply to this line. Please rename the file to something more descriptive than `hardware.py`, for instance `cpu8000.py` though that still doesn't mean much. Maybe include the name of the manufacturer? Also update the class description to something more meaningful and describe the difference with the MCPU800 class. Please also update the class description of the MCPU800 class.

File: python/lsst/ts/tunablelaser/hardware.py [hunk 10]
```diff
@@ -436,12 +412,59 @@ async def update_register(self):
...
         """
-        await self.power_register.read_register_value()
+        await self.power_register.send_command()
 
     def __repr__(self):
         return f"11PMKu:\n {self.power_register}"
 
 
+class MidiOPG:
+    """Implement the MidiOPG module.
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/hardware.py:422]: Please describe what a MidiOPG module is.

File: python/lsst/ts/tunablelaser/hardware.py [hunk 11]
```diff
@@ -436,12 +412,59 @@ async def update_register(self):
...
+        await self.wavelength_register.send_command(value)
+
+    async def update_register(self):
+        """Update all registers."""
+        await self.wavelength_register.send_command()
+
+    def __repr__(self):
+        return f"{self.name}:\n {self.wavelength_register}\n"
+
+
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/hardware.py:467]: Please provide a more descriptive class doc for the MaxiOPG and MLDCO48 classes.

**Review by wvreeven — CHANGES_REQUESTED** [review]
This is a first review of the code changes. I left several comments for you for improvements. When done, please re-request my review and I will take another look.

File: python/lsst/ts/tunablelaser/canbus_modules.py [hunk 12]
```diff
@@ -42,20 +42,25 @@
...
+    "MidiOPG",
 ]
 import logging
 
-from .ascii import AsciiRegister
+from .register import AsciiRegister
 from .enums import Mode, NoSCU, Output, Power, SCUConfiguration
 
 
 class CPU8000:
```
Comment (wvreeven) [inline python/lsst/ts/tunablelaser/canbus_modules.py:53]: It looks like all these laser classes have very similar constructors (both in terms of arguments and functionality) and at least share the `update_register` method. Please extract a common Abstract Base Class (ABC) that sets up the registers and make sure that all these laser classes subclass that ABC.

**Review by wvreeven — CHANGES_REQUESTED** [review]
Looking much better. I found a few more issues that I'd like you to address. Also, the CI build is failing so please fix that (I suspect that it is related to isort).

File: tests/test_ascii.py (hunk 5 above)
Comment (couger01) [reply tests/test_ascii.py]: I fixed the tests.

**Review by wvreeven — APPROVED** [review]
Looks good now. I think the code can still be improved in several ways (like better doc strings) but this has been dragging on already so I'll approve it.

**Pull request merged**